### PR TITLE
[pallas] Added more documentation for grid and BlockSpec.

### PR DESCRIPTION
**Pull request by gnecula**
The starting point was the text in pipelining.md, where I replaced it now with a reference to the separate grid and BlockSpec documentation.

The grids and BlockSpecs are also documented in the quickstart.md, which I mostly left alone because it was good enough for a simple example.

I have also attempted to add a few docstrings.

The rendered documentation is at: https://jax--22134.org.readthedocs.build/en/22134/pallas/index.html and https://jax--22134.org.readthedocs.build/en/22134/jax.experimental.pallas.html#module-jax.experimental.pallas.

### CODE REVIEW

File: jax/_src/pallas/pallas_call.py [hunk 1]
```diff
@@ -1009,6 +1009,41 @@ def pallas_call(
...
+      e.g., as `pl.BlockSpec(lambda *indices: indices, x.shape)`.
+    input_output_aliases: TO BE DOCUMENTED.
+    interpret: runs the ``pallas_call`` as a ``jax.jit`` of a scan over the
+      grid whose body is the kernel lowered as a JAX function. This does not
+      require a TPU or a GPU, and it is useful for debugging.
+    name: TO BE DOCUMENTED.
+    compiler_params: TO BE DOCUMENTED.
+
+  Returns:
+    a function that can be called on a number of positional array arguments to
```
Comment (superbobry) [inline jax/_src/pallas/pallas_call.py:1043]: Nit: uppercase "A".

File: jax/_src/pallas/pallas_call.py [hunk 2]
```diff
@@ -1009,6 +1009,41 @@ def pallas_call(
...
+      The default value for `out_specs` specifies the whole array,
+      e.g., as `pl.BlockSpec(lambda *indices: indices, x.shape)`.
+    input_output_aliases: TO BE DOCUMENTED.
+    interpret: runs the ``pallas_call`` as a ``jax.jit`` of a scan over the
+      grid whose body is the kernel lowered as a JAX function. This does not
+      require a TPU or a GPU, and it is useful for debugging.
+    name: TO BE DOCUMENTED.
+    compiler_params: TO BE DOCUMENTED.
+
+  Returns:
```
Comment (superbobry) [inline jax/_src/pallas/pallas_call.py:1042]: Should we say that the function receiver both input and output refs here or in the documentation of `f` above?

File: jax/experimental/pallas/__init__.py [hunk 3]
```diff
@@ -12,7 +12,10 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-"""Module for pallas, a JAX extension for custom kernels."""
+"""Module for Pallas, a JAX extension for custom kernels.
+
+See the Pallas documentation at https://jax.readthedocs.io/en/latest/pallas.html.
```
Comment (superbobry) [inline jax/experimental/pallas/__init__.py:17]: Can we use a reST reference to link to the Pallas docs instead?

File: docs/pallas/grid_blockspec.md [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+```
+is equivalent to
+```python
+for i in range(n):
+  for j in range(m):
+    # do HBM -> VMEM copies
+    some_kernel(...)
+    # do VMEM -> HBM copies
+```
+This generalizes to any tuple of integers (a length `d` grid will correspond
```
[the remark this answers is not in the document]
Comment (gnecula) [reply docs/pallas/grid_blockspec.md:35]: I think that Pallas allows more than 3 dimensions and will collapse some prefix of the dimensions before lowering to Triton.

File: docs/pallas/grid_blockspec.md [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+On TPUs, programs are executed in a combination of parallel and sequential
+(depending on the architecture) so there are slightly different considerations.
+See [the Pallas TPU documentation](https://jax.readthedocs.io/en/latest/pallas/tpu/details.html#noteworthy-properties-and-restrictions).
+
+(pallas_blockspec)=
+
+### `BlockSpec`, a.k.a. how to chunk up inputs
+
+In conjunction with the `grid` argument, we need to provide Pallas
+the information on how to slice up the input for each grid element.
```
[the remark this answers is not in the document]
Comment (gnecula) [reply docs/pallas/grid_blockspec.md:84]: I will replace it with "for each invocation". I tried to be consistent and refer to invocations, rather than iterations, or blocks.

File: jax/experimental/pallas/__init__.py (hunk 3 above)
Comment (gnecula) [reply jax/experimental/pallas/__init__.py:17]: I was thinking that actual links are helpful for those who read the sources.

File: jax/_src/pallas/pallas_call.py (hunk 2 above)
Comment (gnecula) [reply jax/_src/pallas/pallas_call.py:1042]: Done

File: docs/pallas/grid_blockspec.md [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+Specifically, we need to provide a mapping between *the iteration of the loop*
+to *which block of our inputs and outputs to be operated on*.
+This is provided via {class}`jax.experimental.pallas.BlockSpec` objects.
+
+Before we get into the details of `BlockSpec`s, you may want
+to revisit the
+[Pallas Quickstart BlockSpecs example](https://jax.readthedocs.io/en/latest/pallas/quickstart.html#block-specs-by-example).
+
+`BlockSpec`s are provided to `pallas_call` via the
```
[the remark this answers is not in the document]
Comment (gnecula) [reply docs/pallas/grid_blockspec.md:93]: I added a note that the documentation applies to the `Blocked` mode.

File: docs/pallas/grid_blockspec.md [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+When using {func}`jax.experimental.pallas.pallas_call` the kernel function
+is executed multiple times on different inputs, as specified via the `grid` argument
+to `pallas_call`. Conceptually:
+```python
+pl.pallas_call(some_kernel, grid=(n,))(...)
+```
+maps to
+```python
+for i in range(n):
+  # do HBM -> VMEM copies
```
[the remark this answers is not in the document]
Comment (gnecula) [reply docs/pallas/grid_blockspec.md:17]: I copied this from the `tpu/pipelining.md` :-) I see that the quickstart.md uses SRAM for both TPU and GPU, perhaps that is better?

File: jax/_src/pallas/pallas_call.py [hunk 8]
```diff
@@ -1009,6 +1009,41 @@ def pallas_call(
...
+      See details at :ref:`pallas_grid`.
+    in_specs: a PyTree of :class:`jax.experimental.pallas.BlockSpec` with
+      a structure matching that of the positional arguments.
+      See details at :ref:`pallas_blockspec`.
+    out_specs: a PyTree of :class:`jax.experimental.pallas.BlockSpec` with
+      a structure matching that of the outputs.
+      See details at :ref:`pallas_blockspec`.
+      The default value for `out_specs` specifies the whole array,
+      e.g., as `pl.BlockSpec(lambda *indices: indices, x.shape)`.
+    input_output_aliases: TO BE DOCUMENTED.
```
[the remark this answers is not in the document]
Comment (gnecula) [reply jax/_src/pallas/pallas_call.py:1035]: Done. I left it for later to document how to use it.

File: docs/pallas/grid_blockspec.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (superbobry) [reply docs/pallas/grid_blockspec.md]: I think CUDA calls this SMEM, but that means something else on TPU.

**Review by superbobry — APPROVED** [review]

File: docs/pallas/grid_blockspec.md [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+
+```
+
+We now execute it using `pallas_call` with an additional `grid` argument.
+
+```python
+>>> def iota(len: int):
+...   return pl.pallas_call(iota_kernel,
+...                         out_shape=jax.ShapeDtypeStruct((len,), jnp.int32),
+...                         grid=(len,), interpret=True)()
```
[the remark this answers is not in the document]
Comment (apaszke) [reply docs/pallas/grid_blockspec.md:64]: It's also quite confusing that we use `len` as the parameter, because it shadows the builtin. Perhaps rename to `size`?

File: docs/pallas/grid_blockspec.md [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+>>> iota(8)
+Array([0, 1, 2, 3, 4, 5, 6, 7], dtype=int32)
+
+```
+
+On GPUs, each program is executed in parallel on separate threads.
```
Comment (apaszke) [inline docs/pallas/grid_blockspec.md:70]: Separate blocks actually

File: docs/pallas/grid_blockspec.md [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+(depending on the architecture) so there are slightly different considerations.
+See [the Pallas TPU documentation](https://jax.readthedocs.io/en/latest/pallas/tpu/details.html#noteworthy-properties-and-restrictions).
+
+(pallas_blockspec)=
+
+### `BlockSpec`, a.k.a. how to chunk up inputs
+
+```{note}
+The documentation here applies to the ``indexing_mode == Blocked``, which
+is the default. The documentation for the ``indexing_mode == Unblocked``.
```
Comment (apaszke) [inline docs/pallas/grid_blockspec.md:85]: The last sentence looks unfinished?

File: docs/pallas/grid_blockspec.md [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+
+Informally, the `index_map` of the `BlockSpec` takes as arguments
+the invocation indices (as many as the length of the `grid` tuple),
+and returns **block indices** (one block index for each axis of
+the overall array). Each block index is then multiplied by the
+corresponding axis size from `block_shape`
+to get the actual element index on the corresponding array axis.
+If the resulting element index is such that the block would be
+out of bounds, the element index is reduced so that the entire block
+fits in bounds.
```
Comment (apaszke) [inline docs/pallas/grid_blockspec.md:109]: This sounds a little overly optimistic. The valid block coordinates are those that have their lowest coordinate within data bounds. This implies that blocks with last coordinates along some dimension might contain padding, but it is illegal to access blocks that would contain only padding

File: docs/pallas/grid_blockspec.md [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+ [10 10 10 11 11 11]
+ [20 20 20 21 21 21]
+ [20 20 20 21 21 21]
+ [30 30 30 31 31 31]
+ [30 30 30 31 31 31]]
+
+```
+
+When multiple invocations write to the same elements of the output
+array the result is platform dependent. On the CPU platform the
```
[the remark this answers is not in the document]
Comment (apaszke) [reply docs/pallas/grid_blockspec.md:193]: This is undefined behavior and I don't want to make any promises about it

File: docs/pallas/grid_blockspec.md [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+```python
+pl.pallas_call(some_kernel, grid=(n, m))(...)
+```
+is equivalent to
+```python
+for i in range(n):
+  for j in range(m):
+    # do HBM -> SRAM copies
+    some_kernel(...)
+    # do SRAM -> HBM copies
```
Comment (apaszke) [inline docs/pallas/grid_blockspec.md:33]: I think that the part about HBM SRAM was a good fit for the pipelining tutorial but is not relevant here. Perhaps remove the comments? And perhaps restate this example in the pipelining tutorial with the comments added?

File: docs/pallas/grid_blockspec.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (gnecula) [reply docs/pallas/grid_blockspec.md]: I removed these comments.

File: docs/pallas/grid_blockspec.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (gnecula) [reply docs/pallas/grid_blockspec.md:64]: I renamed to `size`, here and in `quickstart.md` from where I copied this example.

File: docs/pallas/grid_blockspec.md (hunk 14 above)
Comment (gnecula) [reply docs/pallas/grid_blockspec.md:33]: I removed these comments, but I do not want to redo the example in pipelining just for this.

File: docs/pallas/grid_blockspec.md (hunk 10 above)
Comment (gnecula) [reply docs/pallas/grid_blockspec.md:70]: Fixed

File: docs/pallas/grid_blockspec.md (hunk 12 above)
Comment (gnecula) [reply docs/pallas/grid_blockspec.md:109]: Indeed, I did more experiments, and I decided to remove from the documentation for now the handling of out-of-bounds, because there are discrepancies between interpret mode and TPU, and I cannot construct the examples as I want.
